crates/telio-pmtu/src/linux.rs

```diff
@@ -0,0 +1,377 @@
+use std::{ffi::c_int, io, net::IpAddr, os::fd::AsRawFd, time::Duration};
+
+use pnet_packet::{icmp, icmpv6, Packet};
+use telio_utils::telio_log_debug;
+use tokio::net::UdpSocket;
+
+pub struct PMTUSocket {
+    sock: UdpSocket,
+    host: IpAddr,
+}
+
+const TIMEOUT: Duration = Duration::from_secs(1);
+const RETRIES: u32 = 3;
+
+impl PMTUSocket {
+    pub fn new(host: IpAddr) -> io::Result<Self> {
+        let sock = crate::create_icmp_socket(&host)?;
+        set_pmtu_on(&sock, host.is_ipv6())?;
+        Ok(Self { sock, host })
+    }
+    pub async fn probe_pmtu(&self) -> io::Result<u32> {
+        telio_log_debug!("Connecting to: {}", self.host);
+        self.sock.connect((self.host, 0)).await?;
+
+        telio_log_debug!("Connected to: {}", self.host);
+
+        let mut pmtu = read_pmtu(&self.sock, self.host.is_ipv6())?;
+
+        if !super::PMTU_RANGE.contains(&pmtu) {
+            // PMTU is not contained within our reasonable range,
+            // let's probe from the top of the range
+            pmtu = super::PMTU_RANGE.end - 1;
+        }
+
+        telio_log_debug!("Initial PMTU estimate: {pmtu}");
+
+        let mut msgbuf = vec![0xFF; super::PMTU_RANGE.end as usize];
+
+        if self.host.is_ipv6() {
+            let mut pkg = IcmpV6::new(&mut msgbuf)?;
+            self.probe_pmtu_internal(&mut pkg, pmtu).await
+        } else {
+            let mut pkg = IcmpV4::new(&mut msgbuf)?;
+            self.probe_pmtu_internal(&mut pkg, pmtu).await
+        }
+    }
+
+    async fn probe_pmtu_internal(&self, pkg: &mut impl IcmpPkg, mut pmtu: u32) -> io::Result<u32> {
+        pkg.set_ident(super::ICMP_ECHO_REQ_IDENT);
+
+        let mut pmtu_range = super::PMTU_RANGE;
+
+        loop {
+            let check = self.is_pmtu_ok(pkg, pmtu).await?;
+            telio_log_debug!("PMTU {pmtu} result: {check:?}");
+
+            pmtu = match check {
+                PmtuResult::Ok => {
+                    pmtu_range.start = pmtu;
+                    (pmtu_range.start + pmtu_range.end) / 2
+                }
+                PmtuResult::Dropped => {
+                    pmtu_range.end = pmtu;
+                    (pmtu_range.start + pmtu_range.end) / 2
+                }
+                PmtuResult::FragNeeded => {
+                    let mtu = read_error_pmtu(&self.sock, self.host.is_ipv6())?;
```

Comment:
Just wondering. How reliable is it trusting the kernel error message PMTU in favour of our own?
There are huge benefit from our own in this case as it's totally reliable, kernel doesn't really know more than we do.

Reply:
Um, I don't quite get what you mean. There are cases when kernel produces way wrong answer here, though no by its fault. For example when the router responds with `nexthop = 0` then kernel assumes pmtu of 552. Which might not be true. That's why we don't set bound on kernel response but treat it as a hint

Reply:
Sorry for being too ambiguous maybe - I mean you have a manual calculation just above these lines of code doing binary bisect - can we just use that instead and ignore the kernel altogether? This way it's gonna be much simpler :)

Reply:
It might seem so but sometimes it's hard to devise criteria for considering a packet being too big to be transmitted. The kernel reports errors according to its own calculations regardless of whether the packet was transmitted. So it's easier to track down what the kernel does than blindly guess what had happened.
My first implementation was at one point ignoring the kernel reports but still, it was erroring out on `send()` and `recv()` calls sometimes without a reason. The current implementation is more reliable in that sense and is based on the `ping` command code and still, it treats the kernel-tracked pmtu as a hint.
So long story short, it's easier to cooperate with the kernel that fight with it

Reply:
Fair enough, thanks for explaining.

My main concern is about having two solutions where both are not very realiable it seems but look like they do the same thing.

Might I offer adding the reasoning on top for the future readers so they would be confused less?

Reply:
I've added some comments clarifying